sky/provision/kubernetes/instance.py

```diff
@@ -222,6 +222,28 @@ def _wait_for_pods_to_run(namespace, new_nodes):
     Pods may be pulling images or may be in the process of container
     creation.
     """
+
+    def _check_init_containers(pod):
+        # Check if any of the init containers failed
+        # to start. Could be because the init container
+        # command failed or failed to pull image etc.
+        for init_status in pod.status.init_container_statuses:
+            init_terminated = init_status.state.terminated
+            if init_terminated and init_terminated.exit_code != 0:
+                msg = init_terminated.message if (
+                    init_terminated.message) else str(init_terminated)
+                raise config_lib.KubernetesError(
+                    f'Failed to run init container for pod {pod.metadata.name}.'
+                    f' Error details: {msg}.')
+            init_waiting = init_status.state.waiting
+            if (init_waiting is not None and init_waiting.reason
+                    not in ['ContainerCreating', 'PodInitializing']):
```

Comment:
Curious, if `init_waiting.reason` is not in these two states, what else states it could be in and why we directly fail for those cases?

Reply:
Unfortunately there's no listing of state.reason available (it's marked as a [str](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStateWaiting.md) without a list of enum values). Asked an AI assistant and it listed these states:

```
ContainerCreating: The container is still in the process of being created.
CrashLoopBackOff: The container is repeatedly crashing and Kubernetes is backing off before restarting it again.
ErrImagePull: There was an error pulling the container image from the container registry.
ImagePullBackOff: Kubernetes is backing off from pulling the image after a series of failures.
CreateContainerConfigError: There was an error in creating the container configuration.
InvalidImageName: The provided image name is invalid.
PodInitializing: The pod is in the process of initializing.
RunContainerError: There was an error running the container.
ContainerCannotRun: The container cannot run, possibly due to a command or configuration error.
ErrImageNeverPull: The image was not pulled because the policy is set to never pull.
NetworkPluginNotReady: The network plugin is not ready for the container.
BackOff: Kubernetes is backing off from restarting the container.
```

I don't fully trust this list so I don't want to include it in comments, but given this looks like the only successful states are `ContainerCreating` and `PodInitializing`...

Reply:
Should we add a TODO here saying that there might be other states we need to include here when it occurs during usage?

Reply:
Good idea - added now.